src/plot.js

```diff
@@ -84,6 +84,17 @@ export function plot(options = {}) {
   // Initialize the marks’ state.
   for (const mark of marks) {
     if (stateByMark.has(mark)) throw new Error("duplicate mark; each mark must be unique");
+    if (
+      facet?.data &&
+      mark?.data &&
+      mark.data.length === facet.data.length &&
+      mark.data !== facet.data &&
+      facet.data.length > 1
+    ) {
```

Comment:
Here `facet.data` can be an iterable rather than an array (see the call to `arrayify` above), and so `facet.data.length` can be undefined. You can achieve the same result by using `facetIndex.length` instead.

Likewise`mark.data` can be an iterable. This is not converted to an array during the call to _mark_.initialize, but that arrayify’d data is also passed through the mark transform, so we don’t actually store it anywhere. I think this means we’ll need to write a helper function to get the length of the data. We could call `arrayify(mark.data)?.length` here, but that would create a copy of the data in order to perform the test, and it’d be faster not to.

`mark` should never be nullish here, so you don’t need the optional chaining operator to refer to `mark?.data`.

You should use `facets` to check for whether there is faceting as is done elsewhere in this file, rather than `facet?.data`. If `facets` is truthy, then we’re using faceting. (Strictly speaking, it’s possible for faceting to be enabled but for `facet.data` to be truthy, since if you specified `facet: {data: false}` it would be equivalent to `facet: {data: []}` which means that faceting is enabled, but that there are zero facets. That’s because `Array.from(false)` returns `[]`. We could throw an error if there are no facets, I suppose, above, but the main point is that we should be consistent about how we test for faceting, and the best way to be consistent is to check _derived_ state such as `facets` rather than the input options directly, since there are so many defaults and equivalent ways to specify input options.)

Lastly and perhaps most importantly, this isn’t checking the `mark.facet` option: this test should only be applied if the mark.facet option is undefined (or perhaps `"auto"`). Setting this option is how you disable the warning (by removing the ambiguity). The warning message should explicitly instruct you to set that option, for example: “If you intend for this mark to be faceted, set the facet option to true; if you intend for this mark to not be faceted, set the facet option to false.” The warning message should also include the type of the mark so that it’s easier identify, since warning messages don’t have line numbers. I also don’t think the message should refer specifically to filtering as that seems overly specific for what we are testing against. I can provide an alternate suggestion for the wording but I want to focus on getting the logic right first.

Reply:
Wow, thanks for taking the time to walk me through this is such a detailed way. Here's a first pass, but I feel like the check on the `mark.data.size` isn't sufficient:

```js
if (
      facets && // if truthy, faceting is used
      mark.facet === "auto" && // if the user didn't specify a facet
      mark.data &&
      (mark.data.length === facetIndex?.length || 
        mark.data.size === facetIndex?.length)&& // for iterables
      mark.data !== facet.data &&
      facetIndex.length > 1
    )

```

Unsure if there's another check to do without fully copying the `arrayify` function.

Reply:
Unfortunately there’s no easy test for the length of an iterable: in the general case, you actually have to iterate over it. And it’s even further complicated because Array.from doesn’t just allow iterables, it allows “array-likes” like `{length: 10}`. You can look at the [language specification for Array.from](https://tc39.es/ecma262/#sec-array.from) to get a sense of the complexity involved…

We could write an helper that takes data and tests to see if it has the expected length, something like:

```js
function hasLength(data, length) {
  if (data instanceof Array || data instanceof TypedArray) return data.length === length;
  if (data instanceof Map || data instanceof Set) return data.size === length;
  if (typeof data[Symbol.iterator] === "function") {
    let n = 0;
    for (const x of data) {
      if (++n > length) {
        return false;
      }
    }
    return n === length;
  }
  return data.length === length;
}
```

But I think maybe it’s better to just say `arrayify(mark.data).length === facetIndex.length` as the last condition, so that it’s only done if all of the other warning conditions are true.

Reply:
Makes sense, thanks again for clarifying!